### DISCUSSION TITLE
Expression method restrictions

### DISCUSSION THREAD

**Discussion by wojciechsura** [dotnet/csharplang#7089] [Language Ideas]
# Description

Expression methods are a perfect, concise way of writing short methods, which perform a quick data conversion (e.g. mathematical operations). What they miss though is a way to perform some validation on the data beforehand. So, for instance:

```csharp
public int Add(int a, int b) => a + b;

// But

public int Divide(int a, int b) 
{
    if (b == 0)
        throw new ArgumentException(nameof(b));

    return a / b;
}
```

# Proposal

I propose using `where` keyword to extend expression methods, so that they can contain some basic data validation.

```csharp
public int Divide(int a, int b) where (b => b != 0) => a / b;
```

Each condition may be a lambda accepting single parameter and returning bool. Lambda would be in scope of the method declaration, so it would have access to all parameters if necessary (as well as class fields, properties, methods etc.):

```csharp
public int Clamp(int value, int min, int max)
    where (min => min <= max)
    => Math.Max(min, Math.Min(max, value));
```

If condition was not met, an `ArgumentException` would be automatically thrown with lambda parameter as argument name.

# Alternatives

The braced notation is an obvious alternative, though it is not as concise as proposed notation.

Switch expression with pattern matching might be used in cases when only one parameter is validated, eg.

```csharp
public int Divide(int a, int b) => b switch
{
    != 0 => a / b,
    0 => throw new ArgumentException(nameof(b))
};
```

This would get messy though if more parameters required validation (even if it was only a simple one). In such case braced notation would be probably better.

```csharp
// Proposed

public int Process(int a, int b)
    where (a => a > 0)
    where (b => b > 0)
    => a * b;

// Switch expression

public int Process(int a, int b) => a switch 
{
    > 0 => b switch
        {
            > 0 => a * b;
            _ => throw new ArgumentException(nameof(b));
        }
    _ => throw new ArgumentException(nameof(a));
}

// Braced

public int Process(int a, int b)
{
    if (a <= 0)
        throw new ArgumentException(nameof(a));
    if (b <= 0)
        throw new ArgumentException(nameof(b));

    return a * b;
}
```

# Modifications

The lambda validation expression could be shortened even more at the expense of not knowing, which specific argument failed to validate. In such case, I'd propose a new exception to be thrown, e.g. `PreconditionFailedException`.

```csharp
public int Divide(int a, int b) where b != 0 => a / b;

public int Process(int a, int b)
    where (a > 0)
    where (b > 0)
    => a * b;

// Or even simpler

public int Process(int a, int b)
    where (a > 0 && b > 0)
    => a * b;
```

I'm curious if you find this notation useful.

**Comment by HaloFour** [comment]
Seems very similar to method contracts: #105
